File: backend/app/ingestion/chunk.py
"""Page-aware chunking (P1 task #5).

Packs each page's extracted blocks into chunks bounded by
`settings.chunk_max_tokens`, with `settings.chunk_overlap_tokens` of overlap
between consecutive text chunks. Every chunk keeps the page it came from
(OQ-2 page-level citation). Tables are never split: a table block becomes its
own chunk (kept whole for FR-R1), even if it exceeds the size bound.

Token counts here are a whitespace-word approximation — good enough for
sizing. The embedding model (task #6) does the real tokenization.
"""

from __future__ import annotations

import re
from dataclasses import dataclass

from app.config import get_settings
from app.ingestion.extract import PageExtract

_settings = get_settings()
_WORD = re.compile(r"\S+")


@dataclass
class ChunkData:
    page_number: int
    text: str
    token_count: int


def count_tokens(text: str) -> int:
    """Approximate token count (whitespace words)."""
    return len(_WORD.findall(text))


def _words(text: str) -> list[str]:
    return _WORD.findall(text)


def _paragraphs(text: str) -> list[str]:
    """Split a text block into paragraph units (blank-line, else line)."""
    parts = re.split(r"\n\s*\n", text.strip())
    if len(parts) == 1:
        parts = [ln for ln in text.splitlines() if ln.strip()]
    return [p.strip() for p in parts if p.strip()]


def _split_oversized(text: str, max_tok: int, overlap_tok: int) -> list[str]:
    """Split a too-large unit into overlapping word windows."""
    words = _words(text)
    step = max(1, max_tok - overlap_tok)
    windows = []
    for start in range(0, len(words), step):
        windows.append(" ".join(words[start : start + max_tok]))
        if start + max_tok >= len(words):
            break
    return windows


def chunk_pages(
    pages: list[PageExtract],
    max_tokens: int | None = None,
    overlap_tokens: int | None = None,
) -> list[ChunkData]:
    max_tok = max_tokens or _settings.chunk_max_tokens
    overlap_tok = overlap_tokens or _settings.chunk_overlap_tokens
    chunks: list[ChunkData] = []

    for page in pages:
        buf: list[str] = []
        buf_tok = 0

        def flush(page_number: int = page.page_number) -> None:
            nonlocal buf, buf_tok
            if not buf:
                return
            text = "\n\n".join(buf)
            chunks.append(ChunkData(page_number, text, count_tokens(text)))
            # seed next buffer with overlap tail (text continuity only)
            tail = _words(text)[-overlap_tok:] if overlap_tok else []
            buf = [" ".join(tail)] if tail else []
            buf_tok = len(tail)

        for block in page.blocks:
            if block.kind == "table":
                flush()
                buf, buf_tok = [], 0  # no overlap across a table boundary
                chunks.append(
                    ChunkData(
                        page.page_number, block.content, count_tokens(block.content)
                    )
                )
                continue

            for para in _paragraphs(block.content):
                p_tok = count_tokens(para)
                if p_tok > max_tok:
                    flush()
                    buf, buf_tok = [], 0
                    for window in _split_oversized(para, max_tok, overlap_tok):
                        chunks.append(
                            ChunkData(page.page_number, window, count_tokens(window))
                        )
                    continue
                if buf_tok + p_tok > max_tok:
                    flush()
                buf.append(para)
                buf_tok += p_tok

        flush()

    return chunks

File: mobile/app/src/chunk.ts
// Page-aware chunking — direct port of backend/app/ingestion/chunk.py.
//
// Packs each page's blocks into chunks bounded by CHUNK_MAX_TOKENS with
// CHUNK_OVERLAP_TOKENS overlap between consecutive TEXT chunks. Every chunk keeps
// its page_number (OQ-2 page-level citation). Tables are never split: a table
// block becomes its own chunk (FR-R1), even if it exceeds the bound. Token counts
// are a whitespace-word approximation (the embedder does the real tokenization).
//
// Behavior must stay identical to the Python version — verified by the chunk
// parity check (mobile/tools/chunk_parity.mjs).
import { CHUNK_MAX_TOKENS, CHUNK_OVERLAP_TOKENS } from "./config";
import { Block, PageExtract } from "./types";

export interface ChunkData {
  page_number: number;
  text: string;
  token_count: number;
}

// \S+ — runs of non-whitespace (matches Python's re.compile(r"\S+")).
const WORD = /\S+/g;

export function countTokens(text: string): number {
  return (text.match(WORD) ?? []).length;
}

function words(text: string): string[] {
  return text.match(WORD) ?? [];
}

/** Split a text block into paragraph units (blank-line, else line). */
function paragraphs(text: string): string[] {
  let parts = text.trim().split(/\n\s*\n/);
  if (parts.length === 1) {
    parts = text.split(/\r?\n/).filter((ln) => ln.trim());
  }
  return parts.map((p) => p.trim()).filter((p) => p);
}

/** Split a too-large unit into overlapping word windows. */
function splitOversized(text: string, maxTok: number, overlapTok: number): string[] {
  const w = words(text);
  const step = Math.max(1, maxTok - overlapTok);
  const windows: string[] = [];
  for (let start = 0; start < w.length; start += step) {
    windows.push(w.slice(start, start + maxTok).join(" "));
    if (start + maxTok >= w.length) break;
  }
  return windows;
}

export function chunkPages(
  pages: PageExtract[],
  maxTokens: number = CHUNK_MAX_TOKENS,
  overlapTokens: number = CHUNK_OVERLAP_TOKENS,
): ChunkData[] {
  const chunks: ChunkData[] = [];

  for (const page of pages) {
    let buf: string[] = [];
    let bufTok = 0;

    const flush = (): void => {
      if (buf.length === 0) return;
      const text = buf.join("\n\n");
      chunks.push({ page_number: page.page_number, text, token_count: countTokens(text) });
      // Seed next buffer with the overlap tail (text continuity only).
      const tail = overlapTokens ? words(text).slice(-overlapTokens) : [];
      buf = tail.length ? [tail.join(" ")] : [];
      bufTok = tail.length;
    };

    for (const block of page.blocks as Block[]) {
      if (block.kind === "table") {
        flush();
        buf = []; // no overlap across a table boundary
        bufTok = 0;
        chunks.push({
          page_number: page.page_number,
          text: block.content,
          token_count: countTokens(block.content),
        });
        continue;
      }

      for (const para of paragraphs(block.content)) {
        const pTok = countTokens(para);
        if (pTok > maxTokens) {
          flush();
          buf = [];
          bufTok = 0;
          for (const window of splitOversized(para, maxTokens, overlapTokens)) {
            chunks.push({
              page_number: page.page_number,
              text: window,
              token_count: countTokens(window),
            });
          }
          continue;
        }
        if (bufTok + pTok > maxTokens) flush();
        buf.push(para);
        bufTok += pTok;
      }
    }

    flush();
  }

  return chunks;
}
